### DISCUSSION TITLE
Supporting pull-through registry cache/proxy for OCI database

### DISCUSSION THREAD

**Discussion by Patrick-Remy** [aquasecurity/trivy#7817] [Ideas]
### Problem
We are exceeding GitHub rate-limits when using trivy, when (vuln-)db is downloaded via ghcr hosted OCI db, even with correctly configured db-cache in our GitLab CI.

### Cache setup
We are using GitLab CI with configured local filesystem trivy cachedir and caching feature of https://docs.gitlab.com/ee/ci/caching/ 

### Why cache is anyway not really working?
1. GitLab CI/CD caches are seperated by project for security reasons, so even with the same cache-key when pipelines are spawned in different projects the cache can not be shared 
2. We spawn parallel jobs for filesystem and container scanning for performance/time reasons. This is enough that at least (every 12hours?) when cache is invalidated, one of the both parallel trivy jobs exceed GitHub rate limits
3. GitLab CI/CD add suffixes to the cache-keys for protected/unprotected branches by default (which is good for security), so when e.g. publishing a release once a day, trivy always prints that the db need to be updated, as the original cache is a day old, as it is not shared with all the other jobs

### Idea: support registry proxy
As the database is already just an OCI-image, and due to docker-hub rate-limiting many organizations will already have some kind of registry proxy, e.g. via the offical registry image: https://docs.docker.com/docker-hub/mirror/ 

Unfortunately using the existing Parameter [`---db-repository`](https://aquasecurity.github.io/trivy/v0.56/docs/configuration/db/#vulnerability-database) is not working, as the image name needs to be still the same and using `my-regproxy.example.com/ghcr.io/.../trivy-db` is not possible. 

Therefor I'd like to discuss a new parameter like `--db-repository-proxy` to be able to specify the registry proxy url, e.g. `--db-repository-proxy my-regproxy.example.com`.

This would be an extremely easy way without setting up a redis cluster or some similar cache backend to massively reduce the traffic to the used image registries, like GitHub. 

### Target

Container Image

### Scanner

Vulnerability

**Comment by nikpivkin** [comment]
Hi @Patrick-Remy !

Why is `--db-repository` not working for you ? What do you mean by “image name needs to be still the same” ?

**Comment by Patrick-Remy** [comment]
If you setup a pull-through cache with the official docker registry (https://docs.docker.com/docker-hub/mirror/), e.g. reachable under the url `my-regproxy.example.com`, it is not supported to pull an image by using `docker pull my-regproxy.example.com/alpine` or `docker pull my-regproxy.example.com/docker.io/alpine`. The reason is, that this would require you e.g. when building docker images to use `FROM my-regproxy.example.com/alpine` or something similiar in your Dockerfiles. 

Instead when you use this kind of cache/proxy, you just configure your build-daemon, to use a proxy for your images, see: https://docs.docker.com/build/buildkit/configure/#registry-mirror 

And then all image requests to e.g. `docker.io/alpine` are resolved by your proxy/mirror.

**Reply by simar7** [reply]
Maybe I'm not following but how would Trivy use this new flag and determine it's a mirror? Regardless, I am not sure if Trivy should be determining infrastructure parameters. We expect the resolution to be done outside of Trivy so that Trivy simply uses what's provided to it via the `--db-repository` flag.
